### PR TITLE
Add a generic labeled gauge for K8s plugin resources

### PR DESCRIPTION
**Pull request by wild-endeavor**
# TL;DR
Adds a generic labeled gauge for K8s plugin resources

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [x] Smoke tested
 - [x] Unit tests added
 - [x] Code documentation added
 - [x] Any pending items have an associated Issue

## Complete description
* We are adding a generic utility to emit gauge metrics in the `plugin_manager` using the labeled gauges now found in flytestdlib.
* Henceforth, a goroutine will be spun up that will create a `ResourceLevelMonitor` for each type that a plugin registers.
* Changing the existing gauge collecting utilty for FlyteWorkflow CRD objects to use the same labeled gauge.
* We may wish to combine these in the future but as their aggregations are different, we are keeping them separate for now.

## Tracking Issue
https://github.com/lyft/flyte/issues/311

## Follow-up issue
NA

### CODE REVIEW

File: cmd/controller/cmd/root.go [hunk 1]
```diff
@@ -223,6 +223,13 @@ func executeRootCmd(cfg *config2.Config) {
 		logger.Fatalf(ctx, "Failed to start Controller, nil controller received.")
 	}
 
+	go func() {
```
Comment (EngHabu) [inline cmd/controller/cmd/root.go:226]: I merged that change to master... when you rebase you will have it.

File: pkg/controller/nodes/task/k8s/plugin_collector.go [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+	counts := map[string]int{}
+
+	// Collect all workflow metrics
+	for _, v := range objects {
+		metadata, err := meta.Accessor(v)
+		if err != nil {
+			logger.Errorf(ctx, "Error converting obj %v to an Accessor %s\n", v, err)
+			continue
+		}
+		counts[metadata.GetNamespace()]++
```
Comment (EngHabu) [inline pkg/controller/nodes/task/k8s/plugin_collector.go:52]: Worth noting that in L5, everything will look like it was launched in the same namespace because they do not use namespaces to differentiate projects... I _think_ we add project and domain to the labels?

File: pkg/controller/nodes/task/k8s/plugin_collector.go [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+}
+
+func NewResourceLevelMonitor(ctx context.Context, scope promutils.Scope, si cache.SharedIndexInformer, gvk schema.GroupVersionKind) *ResourceLevelMonitor {
+	logger.Infof(ctx, "Launching K8s gauge emitter for kind %s", gvk.Kind)
+	collectorStopWatch := scope.MustNewStopWatchVec("k8s_collection_cycle", "Measures how long it takes to run a collection", time.Millisecond, "kind")
+	return &ResourceLevelMonitor{
+		Scope:          scope,
+		CollectorTimer: collectorStopWatch,
+		levels: scope.MustNewGaugeVec("k8s_resources", "Current Stuff levels",
+			"kind", contextutils.NamespaceKey.String()),
```
Comment (EngHabu) [inline pkg/controller/nodes/task/k8s/plugin_collector.go:109]:
```suggestion
			KindKey, contextutils.NamespaceKey.String()),
```

File: pkg/controller/nodes/task/k8s/plugin_collector.go [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+	go func() {
+		pprof.SetGoroutineLabels(collectorCtx)
+		r.sharedInformer.HasSynced()
+		logger.Infof(ctx, "K8s resource collector %s has synced", r.gvk.Kind)
+		for {
+			select {
+			case <-collectorCtx.Done():
+				return
+			case <-ticker.C:
+				stopwatch, err := r.CollectorTimer.GetMetricWith(map[string]string{"kind": strings.ToLower(r.gvk.Kind)})
```
Comment (EngHabu) [inline pkg/controller/nodes/task/k8s/plugin_collector.go:90]:
```suggestion
				stopwatch, err := r.CollectorTimer.GetMetricWith(map[string]string{KindKey: strings.ToLower(r.gvk.Kind)})
```

File: pkg/controller/nodes/task/k8s/plugin_manager.go [hunk 5]
```diff
@@ -466,11 +470,41 @@ func NewPluginManager(ctx context.Context, iCtx pluginsCore.SetupContext, entry
 		return nil, err
 	}
 
+	// Construct the collector that will emit a gauge indicating current levels of the resource that this K8s plugin operates on
+	var rm *ResourceLevelMonitor
+	rm, err = constructResourceLevelMonitor(ctx, iCtx, metricsScope, entry.ResourceToWatch)
+	if rm != nil {
```
Comment (EngHabu) [inline pkg/controller/nodes/task/k8s/plugin_manager.go:476]: error check?

File: cmd/controller/cmd/root.go (hunk 1 above)
Comment (wild-endeavor) [reply cmd/controller/cmd/root.go:226]: sure will do

File: pkg/controller/nodes/task/k8s/plugin_collector.go (hunk 2 above)
Comment (wild-endeavor) [reply pkg/controller/nodes/task/k8s/plugin_collector.go:52]: but does the spark operator add labels to its pods?

File: pkg/controller/nodes/task/k8s/plugin_collector.go [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+	lock     *sync.Mutex
+	monitors map[schema.GroupVersionKind]*ResourceLevelMonitor
+}
+
+var index monitorIndex
+
+// These are declared here because this constructor will be called more than once, by different K8s resource types (Pods, SparkApps, OtherCrd, etc.)
+// and metric names have to be unique. It felt more reasonable at time of writing to have one metric and have each resource type just be a label
+// rather than one metric per type, but can revisit this down the road.
+var gauge *labeled.Gauge
```
Comment (EngHabu) [inline pkg/controller/nodes/task/k8s/plugin_collector.go:125]: Why not initialize these here?
```suggestion
var gauge = labeled.NewGauge("k8s_resources", "Current levels of K8s objects as seen from their informer caches", scope, labeled.AdditionalLabelsOption{
		Labels: []string{contextutils.NamespaceKey.String(), KindKey.String()},
	})
```

And remove the more complex logic below?

File: pkg/controller/nodes/task/k8s/plugin_collector.go [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+var collectorStopWatch *labeled.StopWatch
+
+func NewResourceLevelMonitor(ctx context.Context, scope promutils.Scope, si cache.SharedIndexInformer, gvk schema.GroupVersionKind) *ResourceLevelMonitor {
```
Comment (EngHabu) [inline pkg/controller/nodes/task/k8s/plugin_collector.go:128]: Nit: maybe rename to `GetOrCreateResourceLevelMonitor` ?

File: pkg/controller/nodes/task/k8s/plugin_collector.go [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+const KindKey contextutils.Key = "kind"
+
+// This object is responsible for emitting metrics that show the current number of a given K8s resource kind, cut by namespace.
+// It needs to be kicked off. The periodicity is not currently configurable because it seems unnecessary. It will also
+// a timer measuring how long it takes to run each measurement cycle.
+type ResourceLevelMonitor struct {
+	Scope promutils.Scope
+
+	// Meta timer - this times each collection cycle to measure how long it takes to collect the levels GaugeVec below
+	CollectorTimer *labeled.StopWatch
```
Comment (kumare3) [inline pkg/controller/nodes/task/k8s/plugin_collector.go:30]: is this a prometheus collector?

File: pkg/controller/nodes/task/k8s/plugin_collector.go [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+	counts := r.countList(ctx, objects)
+
+	for ns, count := range counts {
+		withNamespaceCtx := contextutils.WithNamespace(ctx, ns)
+		r.Levels.Set(withNamespaceCtx, float64(count))
+	}
+}
+
+func (r *ResourceLevelMonitor) RunCollector(ctx context.Context) {
+	ticker := time.NewTicker(resourceLevelMonitorCycleDuration)
```
Comment (kumare3) [inline pkg/controller/nodes/task/k8s/plugin_collector.go:83]: defer ticker.Stop()

File: pkg/controller/nodes/task/k8s/plugin_collector.go (hunk 8 above)
Comment (kumare3) [reply pkg/controller/nodes/task/k8s/plugin_collector.go:30]: if not have you considered using - https://godoc.org/github.com/prometheus/client_golang/prometheus#Collector

Comment (wild-endeavor) [reply pkg/controller/nodes/task/k8s/plugin_collector.go:30]: stopwatches are defined in flytestdlib, but they're basically Summary objects https://github.com/lyft/flytestdlib/blob/c963e9ff8fc9bb46f169c43b923af07d6c5bf71e/promutils/scope.go#L72

File: pkg/controller/nodes/task/k8s/plugin_collector.go (hunk 9 above)
Comment (wild-endeavor) [reply pkg/controller/nodes/task/k8s/plugin_collector.go:83]: you mean inside the goroutine right?  If I add it here, the ticker will just stop and it'll never run.

Comment (wild-endeavor) [reply pkg/controller/nodes/task/k8s/plugin_collector.go:83]: added it inside.

File: pkg/controller/controller.go [hunk 10]
```diff
@@ -271,9 +268,8 @@ func NewResourceLevelMonitor(scope promutils.Scope, lister lister.FlyteWorkflowL
 	return &ResourceLevelMonitor{
 		Scope:          scope,
 		CollectorTimer: scope.MustNewStopWatch("collection_cycle", "Measures how long it takes to run a collection", time.Millisecond),
-		levels: scope.MustNewGaugeVec("flyteworkflow", "Current FlyteWorkflow levels",
-			contextutils.ProjectKey.String(), contextutils.DomainKey.String()),
-		lister: lister,
+		levels:         labeled.NewGauge("flyteworkflow", "Current FlyteWorkflow levels", scope),
```
Comment (kumare3) [inline pkg/controller/controller.go:271]:
```suggestion
		levels:         labeled.NewGauge("flyteworkflow", "Current FlyteWorkflow levels per instance of propeller", scope),
```
